File: ieee/ieee754.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

// Convert float to IEEE-754 binary string
char* ieee754tobin(float num) {
    // Allocate space for 33 characters (32 bits + null terminator)
    char* binStr = (char*)malloc(33);
    if (!binStr) return NULL;

    // Use a union to access float bits
    union {
        float f;
        uint32_t u;
    } converter;

    converter.f = num;
	int i = 0;
    for (i = 31; i >= 0; i--) {
        binStr[31 - i] = ((converter.u >> i) & 1) ? '1' : '0';
    }
    binStr[32] = '\0';
    return binStr;
}

// Convert IEEE-754 binary string to float
float bintoieee754(const char* binStr) {
    if (strlen(binStr) != 32) {
        fprintf(stderr, "Error: Binary string must be 32 bits.\n");
        return 0.0f;
    }

    union {
        uint32_t u;
        float f;
    } converter;

    converter.u = 0;
	int i = 0;
    for (i = 0; i < 32; i++) {
        if (binStr[i] == '1') {
            converter.u |= (1U << (31 - i));
        } else if (binStr[i] != '0') {
            fprintf(stderr, "Error: Invalid character in binary string.\n");
            return 0.0f;
        }
    }

    return converter.f;
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        printf("Usage: %s <float_number>\n", argv[0]);
        return 1;
    }

    float input = atof(argv[1]);

    // Convert float to binary string
    char* binary = ieee754tobin(input);
    if (!binary) {
        fprintf(stderr, "Memory allocation failed.\n");
        return 1;
    }

    printf("IEEE-754 binary of %.6f: %s\n", input, binary);

    // Convert binary string back to float
    float recovered = bintoieee754(binary);
    printf("Recovered float from binary: %.6f\n", recovered);

    free(binary);
    return 0;
}

File: ieee/ieee754.rs
use std::env;
use std::process;

// Convert f32 to IEEE-754 binary string
fn ieee754_to_bin(num: f32) -> String {
    let bits = num.to_bits(); // u32 representation
    let mut bin_str = String::with_capacity(32);
    for i in (0..32).rev() {
        if (bits >> i) & 1 == 1 {
            bin_str.push('1');
        } else {
            bin_str.push('0');
        }
    }
    bin_str
}

// Convert IEEE-754 binary string to f32
fn bin_to_ieee754(bin_str: &str) -> Result<f32, String> {
    if bin_str.len() != 32 {
        return Err("Binary string must be 32 bits.".to_string());
    }

    let mut bits: u32 = 0;
    for (i, ch) in bin_str.chars().enumerate() {
        match ch {
            '1' => bits |= 1 << (31 - i),
            '0' => {}
            _ => return Err("Invalid character in binary string.".to_string()),
        }
    }

    Ok(f32::from_bits(bits))
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <float_number>", args[0]);
        process::exit(1);
    }

    let input: f32 = match args[1].parse() {
        Ok(num) => num,
        Err(_) => {
            eprintln!("Invalid float number.");
            process::exit(1);
        }
    };

    // Convert float to binary
    let binary = ieee754_to_bin(input);
    println!("IEEE-754 binary of {:.6}: {}", input, binary);

    // Convert binary back to float
    match bin_to_ieee754(&binary) {
        Ok(recovered) => println!("Recovered float from binary: {:.6}", recovered),
        Err(err) => {
            eprintln!("Error: {}", err);
            process::exit(1);
        }
    }
}
